curl-helper.c

```diff
@@ -1271,6 +1272,83 @@ static int cb_OPENSOCKETFUNCTION(void *data,
     return ((sock == -1) ? CURL_SOCKET_BAD : sock);
 }
 
+static int cb_OPENSOCKETFUNCTION2(void *data,
+                        curlsocktype purpose,
+                        struct curl_sockaddr *addr)
+{
+    caml_acquire_runtime_system();
+
+    CAMLparam0();
+    CAMLlocal5(result, v_sock_purpose, v_sockdomain, v_socktype, v_sockaddr_record);
+    Connection *conn = (Connection *)data;
+    int sock = -1;
+
+    switch (purpose) {
+        case CURLSOCKTYPE_IPCXN:
+            v_sock_purpose = Val_int(0);
+            break;
+        default:
+            caml_failwith("Invalid socket purpose");
```

Comment:
This function (`cb_OPENSOCKETFUNCTION2`) is called directly by `libcurl`, right? I don't think we want to raise OCaml exceptions here then. Instead we should return `CURL_SOCKET_BAD`.

Reply:
so i understand this argument but there is also another side, the way this can happen is some sort of memory corruption (runtime representation mismatch of ocaml value) - before the change the C frame would be skipped and ocaml would receive an exn and with some luck it will get printed somewhere; with the new code it will be silent failure to download without any hint of what happened. So when the code is in the corruption failed state already i would prefer to get some chance of log mesage instead of silent fail. The rest of callbacks use failwith and it hasn't been reported as a problem so far, another option is to do printf to stderr before cleanup. What do you think?

Reply:
or we can keep as is and introduce an error stream callback to forward error messages from callbacks into ocaml in separate PR as a uniform mechanism for this